tools/perf/util/cs-etm.c

```diff
@@ -514,6 +538,65 @@ static int cs_etm__setup_queues(struct cs_etm_auxtrace *etm)
         return 0;
 }
 
+static inline void cs_etm__copy_last_branch_rb(struct cs_etm_queue *etmq)
+{
+        struct branch_stack *bs_src = etmq->last_branch_rb;
+        struct branch_stack *bs_dst = etmq->last_branch;
+        size_t nr = 0;
+
+        bs_dst->nr = bs_src->nr;
+
+        if (!bs_src->nr)
+                return;
+
+        nr = etmq->etm->synth_opts.last_branch_sz - etmq->last_branch_pos;
+        memcpy(&bs_dst->entries[0],
+               &bs_src->entries[etmq->last_branch_pos],
+               sizeof(struct branch_entry) * nr);
+
+        if (bs_src->nr >= etmq->etm->synth_opts.last_branch_sz) {
+                memcpy(&bs_dst->entries[nr],
+                       &bs_src->entries[0],
+                       sizeof(struct branch_entry) * etmq->last_branch_pos);
+        }
+}
+
+static inline void cs_etm__reset_last_branch_rb(struct cs_etm_queue *etmq)
+{
+        etmq->last_branch_pos = 0;
+        etmq->last_branch_rb->nr = 0;
+}
+
+static void cs_etm__update_last_branch_rb(struct cs_etm_queue *etmq,
+                                          struct cs_etm_packet *packet)
+{
+        struct branch_stack *bs = etmq->last_branch_rb;
+        struct branch_entry *be;
+
+        if (!etmq->last_branch_pos)
+                etmq->last_branch_pos = etmq->etm->synth_opts.last_branch_sz;
+
+        etmq->last_branch_pos -= 1;
+
+        be              = &bs->entries[etmq->last_branch_pos];
+        be->from        = packet->start_addr;
+        be->to          = packet->end_addr;
+        /* No support for mispredict */
+        be->flags.mispred = 0;
+        be->flags.predicted = 1;
+
+        if (bs->nr < etmq->etm->synth_opts.last_branch_sz)
```

Comment:
bs->nr represents the number of branches already pushed to the bs (branch_stack).  It should be incremented until reaching the number of branches the user has specified in the command line "--itrace=il64", i.e., 64.  That number is made available in "etmq->etm->synth_opts.last_branch_sz".
Once bs->nr reaches that number it should remain set to that number: we then read that many entries from the branch stack and generate the last branch event.